helpers/factory: default formatter valid exit code to 0

barrett-ruth opened the pull request:
Resolves #216.

Warns on formatter failures, defaulting to `!= 0`.

Create a utility wrapper `helpers.check_exit_code` to maintain backwards compatibility. 

# TODO

1. `logger:warn` spawns a blocking notification (i.e. must press enter) and creates a timeout error if not pressed within the defualt timeout

^ this is unavoidable - addressed by using a shorter initial message pointing you to to the log file.

barrett-ruth commented:
@mochaaP I can't assign reviewers so please take a look. I know I need to squash commits and a bunch of other stuff but I just wanted to draft the initial API. Let me know everything that can be improved.

barrett-ruth commented:
@mochaaP I'm going to start going through the formatters one-by-one (pain) to ensure I'm passing the right exit code. 

1. Is there a better way to do this?
2. I currently make it optional (and default to exit code 0) to indicate formatter success in `check_exit_code`. Is this fine?

mochaaP commented:
I think (2) is fine.

For (1), we could land the changes first and wait for feedback from ones who actually use them if any problem arises.

barrett-ruth commented:
Ready for review.

mochaaP commented on lua/null-ls/helpers/formatter_factory.lua:
```diff
@@ -18,5 +18,10 @@ return function(opts)
     end
 
+    if opts.check_exit_code == nil then
+        vim.print(opts)
```
Forgot to clean up a debug print?

mochaaP commented on lua/null-ls/helpers/check_exit_code.lua:
```diff
@@ -0,0 +1,20 @@
+local logger = require("null-ls.logger")
+
+return function(success_codes, command)
+    return function(code, stderr)
+        local success
+
+        if type(success_codes) == "number" then
+            success = code <= success_codes
+        else
+            success = vim.tbl_contains(success_codes, code)
+        end
+
+        if not success then
+            vim.schedule(function()
+                logger:warn(string.format("failed to run formatter %s; see `:NullLsLog`", command))
+                logger:add_entry(string.format("failed to run formatter %s: %s", command, stderr), "warn")
+            end)
+        end
+    end
+end
```
duplicated here:

https://github.com/nvimtools/none-ls.nvim/blob/main/lua%2Fnull-ls%2Fhelpers%2Fgenerator_factory.lua#L135-L140

mochaaP commented on lua/null-ls/helpers/formatter_factory.lua:
```diff
@@ -18,5 +18,10 @@ return function(opts)
         return done({ { text = output } })
     end
 
+    if opts.check_exit_code == nil then
```
Set `check_exit_code` in the generator factory instead of here.

barrett-ruth commented:
@mochaaP addressing comments above but this is a longer response:

The problem I have with moving the code to the generic factory (and this includes the code duplication comment above) is that I thought this feature only makes sense for formatters.

If we added this to generator factory, wouldn't we also be checking linting exit codes as well?

barrett-ruth commented:
@mochaaP this is ready for review, see above comments

barrett-ruth commented:
What does the eyes emoji mean :sob: I've already addressed the comments I thought... what do you think? (lol)

mochaaP commented:
“I will take a look at it” 😂 This needs some slight changes, I will take care of it.

mochaaP commented:
Do you think this implementation is okay?

barrett-ruth commented:
I see. You just set the exit code to `0` in the code and then that's picked up later in the parsing.

Much cleaner implementation, LGTM.

barrett-ruth commented:
@mochaaP you updated the logging message to use `logger:add_entry`.

Then you deleted the file containing that change (`check_exit_code`) - lol! I'm wondering if this was intentional? I thought we wanted this informative message to the end user?

If this is the case, I thought we had to separate the logic out in another file to do that log... because *only* in the formatters does the notion of validating the exit code make sense.

Sorry this is taking so long, LMK if you know what I'm saying.

mochaaP commented:
ack, will check tonight.

barrett-ruth commented:
LGTM!

geril2207 commented:
@mochaaP @barrett-ruth 
prettierd nvim(nightly but idk if it changes anything):

https://github.com/user-attachments/assets/32938e71-b5d5-430f-ba95-b89c35eaacea

before:

https://github.com/user-attachments/assets/ccdb13ba-d102-472c-a25f-19be9b010b9e

mochaaP commented:
cc @barrett-ruth is prettierd outputting to stdout? we might need to ignore that

daephx commented:
Found myself here because I updated none-ls and keep getting the error `nvim_echo must not be called in a lua loop callback` when the buffer has invalid syntax. I've only tested with `stylua`, and haven't checked other formatters.

The error output itself is rather annoying, so for now my solution is to pin none-ls to commit 83c7c00.

mochaaP commented:
I've partially reverted this until I figure out a better way to present the error message.
